stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -1507,6 +1512,58 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
   }
 }
 
+/// A TensorFlow device kind.
+public enum DeviceKind : UInt {
+  /// Default device.
+  case `default` = 0
+  /// Central processing units.
+  case cpu = 1
+  /// Graphics processing units.
+  case gpu = 2
+  // TODO: TPU?
+}
+
+internal extension _ExecutionContext {
+  @usableFromInline
+  func pushDeviceScope(_ kind: DeviceKind) {
+    sync {
+      deviceScopes.append(kind)
+    }
+  }
+
+  @usableFromInline
+  func popDeviceScope() {
+    sync {
+      internalConsistencyCheck(deviceScopes.popLast() != nil)
+    }
+  }
+}
+
+/// Execute ops in a provided closure on a specified device.
+///
+/// - Parameters:
+///   - kind: The device to run the ops on.
+///   - body: The function to run.
+@inlinable
+public func withDevice<R>(_ kind: DeviceKind, perform body: () -> R) -> R {
+  _ExecutionContext.global.pushDeviceScope(kind)
+  let result = body()
+  _ExecutionContext.global.popDeviceScope()
+  return result
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetDeviceFromScope")
+func _TFCOpSetDeviceFromScope(_ op: CTFEOp, _ status: CTFStatus) {
+  let context = _ExecutionContext.global
+  let device = context.sync { context.deviceScopes.last ?? .`default` }.opSetDeviceFromScope(op, status)
```

Comment:
> context.deviceScopes.last ?? .`default`

Given we initialized `deviceScopes` with a root element (be it `.cpu`, or `.default` as i suggested above), should we instead assert that `context.deviceScopes.last` is never `'nil`?